### ISSUE TITLE
[Bug]: Whisper segment timestamps collapse under `prompt` — the reference timestamp decoding rules are not applied

### ISSUE THREAD

**Issue by hanslin903** [vllm-project/vllm#55281]
### Your current environment

- vLLM `v0.28.0` (official image `vllm/vllm-openai:v0.28.0`, plus `librosa soundfile av`)
- Model: `openai/whisper-large-v3-turbo`, `--enforce-eager --max-num-seqs 4`
- Endpoint: `POST /v1/audio/transcriptions`, `response_format=verbose_json`
- **The fix from #53207 (`<|startofprev|>`) was applied for every measurement below.** Without it the `prompt` field is broken for an unrelated reason and the numbers here cannot be reproduced.

### 🐛 Describe the bug

**Passing `prompt` to a `verbose_json` transcription collapses the segmentation.** The text survives; the timestamps do not. A 30 s clip that yields 9 segments without a prompt yields 1 segment with one, spanning the whole window.

This is not the `<|prev|>` bug (#53206) — that is fixed in these runs.

#### Measurements

30 s Mandarin public-service announcement, `language=zh`, `temperature=0`, deterministic across repeats. Prompt is a list of domain terms (the usual "hotwords" use case), lengthened word by word:

| `prompt` | prompt tokens | segments | chars | last segment ends |
|---|---|---|---|---|
| *(none)* | 0 | **9** | 165 | 30.4 |
| 4 words | 14 | 8 | 123 | 29.8 |
| 8 words | 29 | **2** | 140 | 30.4 |
| 12 words | 44 | **1** | 132 | 29.8 |
| 16 words | 56 | **2** | 149 | 30.4 |
| 20 words | 76 | 1 | 15 | *(output degenerated to `ご視聴ありがとうございました。`)* |
| 24 words | 87 | **0** | 0 | — |
| 30 words | 101 | **0** | 0 | — |

The prompt never approaches the decoder budget (101 tokens out of `max_model_len=448`), and the transcript itself stays complete up to 16 words — only the segmentation degrades.

#### The timestamps are not being emitted at all

`verbose_json` returns each segment's `tokens`. Timestamp tokens for this model start at `<|0.00|>` = **50365**. Counting them inside the returned segments:

```
no prompt   : 9 segments, 0 timestamp tokens inside any segment  (all at the boundaries, as expected)
12-word prompt : 1 segment, 139 tokens, 0 timestamp tokens inside
```

So the model transcribes the full 30 s but emits **no intermediate timestamps whatsoever**. Nothing is being lost by the parser — the information never exists.

#### The same weights do not behave this way under the reference implementation

`faster-whisper` 1.2.1 (CTranslate2 4.8.2), same checkpoint, same audio, same word lists.
vLLM ran fp16 on GPU, faster-whisper int8 on CPU — that difference shows up in the
transcribed text, not in whether timestamps are emitted:

| `prompt` | vLLM segments | faster-whisper segments |
|---|---|---|
| *(none)* | 9 | 8 |
| 4 words | 8 | 7 |
| 8 words | **2** | 6 |
| 12 words | **1** | 6 |
| 16 words | **2** | 7 |
| 20 words | **1** | 7 |
| 24 words | **0** | 3 |
| 30 words | **0** | 6 |

### Where it comes from

vLLM decodes Whisper as a generic encoder-decoder LM. The reference implementation does not: it constrains the logits at every step with `ApplyTimestampRules` (`whisper/decoding.py`), which

- suppresses `<|notimestamps|>` when timestamps are requested,
- forces timestamps to appear **in pairs** and to increase monotonically,
- requires the sequence to begin with a timestamp,
- bounds the first timestamp by `max_initial_timestamp`.

CTranslate2 implements the same rules inside its Whisper-specific `generate()`, which is why `faster-whisper` can expose `max_initial_timestamp` as an option at all.

vLLM has none of this. Instead, `_preprocess_verbose_prompt` in `vllm/entrypoints/speech_to_text/base/serving.py` string-replaces `<|notimestamps|>` with `<|0.00|>` in the decoder prefix:

```python
dec_prompt["prompt"] = dec_prompt["prompt"].replace("<|notimestamps|>", "<|0.00|>")
```

This is a stand-in for the missing rules: it guarantees the output *starts* as if in timestamp mode, and nothing after that keeps it there. A prompt — plain prose carrying no timestamp tokens — biases the decoder toward continuous text, and with no constraint to emit paired timestamps it simply stops emitting them.

### The obvious fix does not work — I tried it

Removing the replacement (so the model generates its own first timestamp, as the reference does) makes things worse, not better:

| `prompt` | segments before | segments after |
|---|---|---|
| *(none)* | 9 | **0** |
| 4 words | 8 | 8 |
| 8 words | 2 | **9** |
| 12 words | 1 | 1 |
| 16 words | 2 | 1 |

Logging the first generated token explains it:

```
no prompt      -> 50364  <|notimestamps|>   (the model opts out of timestamps entirely)
4-word prompt  -> 50365  <|0.00|>
12-word prompt -> 50365  <|0.00|>           (in timestamp mode, still only 1 segment)
```

Without `ApplyTimestampRules` suppressing it, `<|notimestamps|>` is simply a legal continuation and the model takes it. And the 12-word row shows that even entering timestamp mode is not enough: with no pairing or monotonicity constraint, the decoder can run to the end of the window without emitting another timestamp.

So the prefix is not the cause, and the prefix hack cannot be removed on its own.

### Suggested fix

Implement Whisper's timestamp rules as a logits processor applied to transcription requests when segment timestamps are requested — the four constraints listed above. That would let `_preprocess_verbose_prompt` drop the string replacement, and would make segment timestamps hold up under `prompt`.

This also matters for #53144: advancing the cursor to the end of the last segment relies on the timestamps being emitted correctly in the first place, which today is not guaranteed.

### Related

- #53206 / #53207 — `<|prev|>` prompt token; fixed in all measurements above
- #53144 — audio dropped when the decoder stops early; complementary, and depends on this
- #47664 — word-level timestamps
- #25750 — Whisper feature tracking

### Before submitting a new issue...

- [x] Made sure I searched the documentation and existing issues.

**Comment by Rishabh2561** [comment]
One additional reference behavior may be important for the proposed fix: `ApplyTimestampRules` includes a probability-mass rule beyond the four structural constraints listed above.

After applying the timestamp masks, Whisper computes the combined log-probability of all timestamp tokens. If that exceeds the maximum log-probability of any individual text token, it suppresses the text-token range and forces a timestamp:

```python
timestamp_logprob = logprobs[timestamp_begin:].logsumexp(dim=-1)
max_text_token_logprob = logprobs[:timestamp_begin].max()
if timestamp_logprob > max_text_token_logprob:
    logits[:timestamp_begin] = -inf
```

This seems especially relevant when `prompt` biases the decoder toward continuous text. Implementing only suppression, pairing, monotonicity, and initial-timestamp bounds might still diverge from Whisper's decoding behavior.

For vLLM, this could fit as a request-aware `AdapterLogitsProcessor`, enabled only for Whisper segment-timestamp requests through `SamplingParams.extra_args`. Because that interface receives generated `output_token_ids` separately from the decoder prompt, the timestamp state machine can ignore `<|startofprev|>` and prompt tokens naturally.

Useful unit cases would cover:

- the aggregate timestamp-probability rule;
- first-token and `max_initial_timestamp` bounds;
- timestamp pairing and nonzero segment lengths;
- monotonic timestamp enforcement;
- prompt tokens being excluded from timestamp state;
- independent state for multiple requests in a batch.

Reference: https://github.com/openai/whisper/blob/main/whisper/decoding.py#L441-L462
